Rotary encoder additional invert options for B&W screens (replacement)

breadoven opened the pull request:
Replacement for damaged PR https://github.com/EdgeTX/edgetx/pull/1826.

Adds 2 additional rotary encoder invert options for B&W screens changing behaviour of menu navigation including popup menus. Not used for colour screens due to different navigation method.

Rotary Invert setting changed from Check box to Combo with Normal, Inverted, V-N, V-A where:

Normal = vertical and horizontal "normal"
Inverted = vertical and horizontal inverted
V-N = vertical movement inverted (Normal), horizontal row movement behaviour unchanged (always cycles through all settings in rows)
V-A = vertical movement inverted (Alternative), horizontal row movement behaviour changed so left scroll moves selection directly down LH settings without cycling through multiple settings in rows. Movement along rows of settings only happens when right scrolling, moving up a row when scrolling beyond last setting in row.

LUA API function added providing rotary encoder inverted status.

Sim tested OK on X9D+ 2019, ZORRO, TX16S, TPRO, T18, TX12, T12. QX7 ACCESS, X10 EXPRESS. Fully tested on ZORRO radio.

gagarinlg commented on radio/src/gui/common/stdlcd/popups.cpp:
```diff
@@ -94,7 +94,19 @@ const char * runPopupMenu(event_t event)
+  event_t eventTemp = event;
+
+#if defined(ROTARY_ENCODER_NAVIGATION) && !defined(COLORLCD)
+  if (g_eeGeneral.rotEncDirection >= 2) {
```
Please make defines for the different rotary encoder modes and ause then at all places where rotEncDirection is checked.

pfeerick replied:
Done as enums.

gagarinlg commented on radio/src/lua/api_general.cpp:
```diff
@@ -553,6 +553,26 @@ static int luaGetRotEncSpeed(lua_State * L)
   return 1;
 }
 
+/*luadoc
+@function getRotEncInvert()
+
+Return rotary encoder inverted status
+
+@retval number in list: OFF = 0, ON = 1, V-N = 2, V-A = 3
+  return 0 on radio without rotary encoder
+
+@status current Introduced in 2.8.0
+*/
+static int luaGetRotEncInvert(lua_State * L)
+{
+#if defined(ROTARY_ENCODER_NAVIGATION)
+  lua_pushunsigned(L, g_eeGeneral.rotEncDirection);
+#else
+  lua_pushunsigned(L, 0);
+#endif
+  return 1;
+}
+
```
is "...Invert..." the right name? Maybe "getRotEncMode" would be better.

pfeerick replied:
Since this now describes more than rotary encoder inversion, yes... it's now getRotEncMode

gagarinlg commented on radio/src/translations.cpp:
```diff
@@ -53,6 +53,7 @@ ISTR(VTELEMSCREENTYPE);
 ISTR(VKEYS);
 ISTR(VSWITCHES);
 ISTR(VSRCRAW);
+ISTR(INVERT_ROTARY_OPT);
```
Same as before, rethink the name. Maybe "ROTARY_MODE_OPT"?

gagarinlg commented on companion/src/firmwares/generalsettings.h:
```diff
@@ -231,7 +231,7 @@ class GeneralSettings {
     bool rtcCheckDisable;
     bool keysBacklight;
-    bool rotEncDirection;
+    unsigned int rotEncDirection;
```
Please rename to "rotEncMode", or, since the last word gets shorter, "rotEncoderMode"

gagarinlg reviewed (changes requested):
All the whitespace removel changes are nice as they clean up the files, but they really clutter this PR.

elecpower commented:
Companion side looks okay

pfeerick commented:
@elecpower Damn ... did you have your psychic hat on? That question was going to be asked soon... 😆

pfeerick commented:
Whoops... just missed the nightly window... looks like this is ready to merge though. Just a last minute string change and changing the UI on colorlcd so that it is a choice there also (which while not only consistent, also means it's ready if more options are added on the colorlcd side also).

breadoven commented:
Thanks for finishing this off, easier for you than me having to work out all the changes since last time I looked at it.

Just need to sort out the Telemetry button now so it does more than just turn on the telemetry page ( (assuming not already done).

pfeerick commented:
I don't things there's been any change there, so all yours ;)
